[WIP] First draft of Workflow documentation

sigmavirus24 opened the pull request:
As discussed this morning on our US/UK developer meeting, here is the
documentation around the current workflow and a workflow based on what
we discussed this morning.

---

**Note** I pushed this to rpc-openstack so others could update it as necessary. This is in accordance with the proposed workflow in the document.

**Please** add yourself to the commit message as a co-author if you update this.

nrb commented on WORKFLOW.rst:
```diff
@@ -0,0 +1,153 @@
+   fix to a branch on |rpcos|. [#]_
+
+#. A PR is sent to target the |lsbnext| branch (if the bug applies to
+   |lsbnext|, otherwise it is sent to the most recent branch it affects.
```
In this workflow, when would the `liberty` branch be created?

sigmavirus24 replied:
Notice that I specifically avoided talking about releases. That is something I think we didn't agree upon in the meeting so I wasn't going to try to codify something that would start an ideology war. We can iterate on this document to add that in either in this pull request or in a later pull request.

sigmavirus24 replied:
So now that I've added something, I think the workflow I'd like to propose (that's closest to what we have) is something along the lines of the following:

Currently we're still developing things for the 11.x series (kilo). Master would then be named `kilo-next`. When we decide to start doing liberty development, `kilo-next` would become `liberty-next`. When we create r12.0.0 is when we would create a `liberty` branch. That said, if we change our workflow to the strawman below, we would then simply cut `r12.0` branch from `liberty-next` instead.

stevelle replied:
To be clear:
Are you suggesting that we would also possibly create a `kilo-next` branch when the default branch is renamed to `liberty-next` (because we have adopted and backported major features, like https://review.openstack.org/#/c/244215/ before and are likely to do it again).

I'm not clear yet what role `liberty` plays as opposed to `liberty-next` either.

I think we need several examples of the branching behaviors with visuals to be sure. That might also make it easier to spot gaps in this outline. I started sketching but realized I don't understand how to put these pieces together yet in your model, and would need to invent my own model to complete them.

sigmavirus24 replied:
```
+--------+------------------------------+--> kilo-next+-------------+-----------> liberty-next
         |                              |                           |
         +------------------> r11.0     |                           |
                                        +---------> r11.1           |
                                                                    |
                                                                    +--------->r12.0
````

Where each minor branch is created *at release time*

sigmavirus24 replied:
```
       +
       |
       |
       |
       +-------------------+
       |                   |
       |                   |
       +-------------+     |
       v             |     v
   kilo|next         |  r11.0
       |             |
       |             v
       |          r11.1
       |
       |
       +--------+
       v        |
liberty-next    |
                v
             r12.0
```

Might be easier to read

sigmavirus24 commented:
On second thought, I'm going to add separate sub-sections for release workflow (current, and a straw man proposed one).

sigmavirus24 commented:
So in d0bb4f3 I updated subsection headings and added descriptions of the current release workflow. A rendered view of the document can be seen [here](https://github.com/rcbops/rpc-openstack/pull/667/files?short_path=bb33e90#diff-bb33e90f50a7404827488e9759e50e7d).

git-harry commented on WORKFLOW.rst:
```diff
@@ -0,0 +1,216 @@
+
+#. Bug fixes move from ``master`` to the other branch(es) via cherry-picks.
+
+#. Bugs are hand-managed by developers and backports are tracked haphazardly.
```
What does 'hand-managed' refer to?

'haphazardly' feels more like an opinion than a fact.

sigmavirus24 replied:
All bugs that are addressed by changes have to be hand-closed by someone. There is no automation.

"haphazardly" is accurate and factual. We currently rely on people noticing that all backports have been merged before anything is closed. This (I believe) is skewing our Kanban metrics against us and is far more work than we should be doing. GitHub has tooling for closing issues for us, we should be utilizing that, not working against it.

sigmavirus24 replied:
A perfect example of what I'm talking about here is #525. This could have been closed a long time ago, but because of our existing workflow, we only just closed it seconds ago.

git-harry commented on WORKFLOW.rst:
```diff
@@ -0,0 +1,216 @@
+
+Release Workflow
+----------------
+
```
There is no mention of patch releases in this section.

sigmavirus24 replied:
Good point.

git-harry commented on WORKFLOW.rst:
```diff
@@ -0,0 +1,216 @@
+
+.. note::
+
+    When a new stable release of OpenStack is created, |lsbnext| would then be
```
Is 'stable release of OpenStack' referring to the newest stable release of rpcos?

git-harry replied:
This seems to be suggesting branching early. So relative to the current branch setup, the only difference would be that master has a different name. I'd prefer to branch when work starts on the next major release and not when a new stable release is created unless we intend to maintain stable minor releases.

sigmavirus24 replied:
It's not yet clear (as this is currently worded) but what I'm trending towards, personally, is https://github.com/rcbops/rpc-openstack/pull/667/files#r47525797 but I want feedback on all of this.

git-harry commented on WORKFLOW.rst:
```diff
@@ -0,0 +1,216 @@
+
+    When a new stable release of OpenStack is created, |lsbnext| would then be
+    renamed to the codename of that release, e.g., ``liberty-next``.
+
```
What benefit does this provide over leaving it called master?

sigmavirus24 replied:
Explicit is far better than implicit. `master` is a meaningless, generic term. `blah-next` tells people the branch's purpose better than `master`. We've had a fair number of people confused about what `master` was since we released `r11.0.0`. Ideally this naming change makes the repository more self-explanatory.

git-harry replied:
Agreed about the benefit of being explicit. Not necessarily against it. I wonder if this has any sort of knock on effects, for example anything testing master will now need to be periodically updated to reference the new branch.

git-harry commented on WORKFLOW.rst:
```diff
@@ -0,0 +1,216 @@
+       - [ ] ...
+
+   Using markdown checklists in GitHub will give others a quick view into the
+   status of each pull request.
```
Tags are nice because they all lists of issues to be filtered.

sigmavirus24 replied:
Huh?

sigmavirus24 replied:
Are you thinking about Milestones? Or maybe Release objects?

git-harry replied:
Maybe I misunderstood this section. As I read it the issue description would be used to track the branches to which the issues pertains. Currently tags are used to flag and issues for backports and that is useful be issues can be filtered based on a tag.

git-harry commented on WORKFLOW.rst:
```diff
@@ -0,0 +1,216 @@
+    practices.
+
+#. Each time a large feature targetted for the next minor release is merged, a
+   Beta tag is created, e.g., ``r11.1.0b1``, ``r11.1.0b2``, etc.
```
What is the benefit of beta tags?

sigmavirus24 replied:
Tagging early gives QE and support something they can check out without having to document/pass around/memorize a SHA. More tags ideally leads to shorter feedback loops.

git-harry commented on WORKFLOW.rst:
```diff
@@ -0,0 +1,216 @@
+#. As we approach the end of features and bugs targetted for a minor release,
+   we create a Release Candidate tag, e.g., ``r11.1.0rc1``.
+
+#. Each tag that is created should be tested in depth by QE and/or Support for
```
What is the frequency of new tags to test?

sigmavirus24 replied:
Anything we decide on too rigidly will not be followed. We don't want to overload QE & Support, but we should give them some kind of milestone that they can use to provide earlier feedback.

git-harry replied:
Agreed. I'm just trying to understand what extra work is implied by these tags and whether it would require additional resources.

git-harry commented on WORKFLOW.rst:
```diff
@@ -0,0 +1,216 @@
+#. The developer submits a pull request against ``master``. It's reviewed,
+   iterated upon, and eventually merged.
+
+#. If that PR is elligible to be backported, then the associated bug is
```
s/elligible/eligible/

git-harry commented:
In terms of the tools we use, I would still prefer we look to switch to launchpad/gerrit. While Github is perfectly adequate for our needs I would prefer to consolidate the tools that are used. The majority of the work done on this project and any others the team hosts on Github is by us. It strikes me that it would be easier for us if we need to be familiar with less tools and that would also make it easier for others to get started.

Regarding the workflow in terms of when/how we branch/tag etc, I think that needs to be informed by how support actually use the software. There is no point trying to force a process on support that we will never get them to consume.

stevelle commented on WORKFLOW.rst:
```diff
@@ -0,0 +1,216 @@
+.. Replacements
+
+.. |rpcos| replace:: ``rpc-openstack``
+.. |lsbnext| replace:: ``{{latest-stable-branch-name}}-next``
```
latest-stable-branch-name is not as descriptive as I would like, as then we have to explain that a branch is stable when... and then have throw down qualifiers around work to try and get a next-cycle release together.

sigmavirus24 replied:
I'm open to alternative names. Perhaps "most recent release series codename"?

stevelle replied:
that might work, but it is so long you can lose the context by the time you reach the end of the variable name... maybe just {{ series-codename }}-next ?

I think *anything* we put here may need to be decoded / explained so maybe we accept that and move on. I'm not sure, and this is a meta-concern about the usability of this document, not the actual process it is modeling so not critical.

stevelle commented on WORKFLOW.rst:
```diff
@@ -0,0 +1,216 @@
+
+#. When a release is created, branches can be created from |lsbnext| if
+   desired by the stakeholders.
+
```
Needs some mention of how we would deal with Next Major.

sigmavirus24 commented:
> In terms of the tools we use, I would still prefer we look to switch to launchpad/gerrit. While Github is perfectly adequate for our needs I would prefer to consolidate the tools that are used. The majority of the work done on this project and any others the team hosts on Github is by us. It strikes me that it would be easier for us if we need to be familiar with less tools and that would also make it easier for others to get started.

Then we need some solid reasons as to why. If GitHub is perfectly adequate, why do we have to change to launchpad and gerrit?

Which gerrit should we be using? review.openstack.org? I think having everything there is going to be convenient for us and us only. We've had an increased participation (in terms of contributions) from our support teams and other teams (e.g., @BjoernT, @byronmccollum, @dpham-rs, etc.) and I can think of only one that I've seen contribute to OSA on review.o.o. Do we really want to change that out from under them as they're just beginning to become comfortable contributing? I don't.

We have already introduced a broken process for how we develop on GitHub. I'd like to see us try fixing the process instead of throwing everything out and using a new system that will alienate people who are just beginning to contribute.

Further, we already have sufficient evidence that the LaunchPad `<->` review.o.o integration is flakey and unreliable at best. It's not native integration. It does not currently support our (apparently new release strategy where we have `kilo`, `kilo-r11.1.0` (and it most certainly will not, because it was a fight to get the jeepyb change in for OSA in the first place). So then we'll have to maintain our own integration or do more work by hand? Do we really want that? Why?

But I'll pose the question again (because I'd like an answer), "Why is launchpad + 'gerrit' better?"

git-harry commented:
@sigmavirus24 at least two of the names you list have used launchpad/gerrit. Is you're description of their level of comfort based on some form of survey or is it conjecture?

I don't remember seeing this apparent new release strategy announced anywhere, apart from the fact that I'd like to understand why it was changed - it can always be changed again.

The crux of it is that everyone contributing to OpenStack needs familiarity with launchpad/gerrit so why not just piggyback on that instead of having a completely separate tool chain for rpc-os when it provides us with no obvious benefit. By synchronising it even reducing the barrier to entry for support getting involved with OpenStack and seeing that we provide OpenStack training there will likely be a whole load of existing material that can be used to assist support instead of the training team needing to produce something new.

sigmavirus24 commented:
> Is you're description of their level of comfort based on some form of survey or is it conjecture?

First, that list is not complete. You know that, arguing based on an example list of names is not worth anyone's time. Conjecture based on my own perceived level of activity on both trackers from several people not directly on our team.

> I don't remember seeing this apparent new release strategy announced anywhere, apart from the fact that I'd like to understand why it was changed - it can always be changed again.

It wasn't announced. It was a casualty of our "official" early branching strategy that made the `kilo` branch something we could not cleanly or easily merge master into (a workflow that Gerrit doesn't support).

> The crux of it is that everyone contributing to OpenStack needs familiarity with launchpad/gerrit so why not just piggyback on that instead of having a completely separate tool chain for rpc-os when it provides us with no obvious benefit.

Why do they need familiarity with it though? How much do they actually contribute to Nova, Neutron, Glance, etc.?

Why is there no obvious benefit from using a service (like GitHub) that has tighter and far more reliable integrations than review.o.o + launchpad?

> seeing that we provide OpenStack training there will likely be a whole load of existing material that can be used to assist support instead of the training team needing to produce something new.

Our OpenStack training does not teach you how to contribute. It teaches you how to stand-up a cloud and use it. Also, presently, we don't need training for this project because there's sufficient training documentation provided by GitHub and the internet because it's such a common workflow.

If you're anticipating our other teams making significant contributions to OpenStack services, they'll also be making significant contributions to OpenStack's dependencies as well which are hosted on:

- GitHub
- BitBucket
- GitLab
- And many other places

If your only argument is that switching to those services and tools will provide them with the training they need to do this thing that they have in the past adamantly proclaimed as not their job, we shouldn't only be looking at review.o.o (gerrit) + launchpad.

byronmccollum commented:
> Our OpenStack training does not teach you how to contribute.

We _**do**_ teach the OS contributing toolchain...

:sunglasses:

sigmavirus24 commented:
> We do teach the OS contributing toolchain...

This is new then (at least it wasn't around when Dewey and a bunch of us went through it). Awesome!

git-harry commented:
>> Is you're description of their level of comfort based on some form of survey or is it conjecture?

>First, that list is not complete. You know that, arguing based on an example list of names is not worth anyone's time. Conjecture based on my own perceived level of activity on both trackers from several people not directly on our team.

I wasn't claiming the list to be complete, I was simply trying to show that experience of launchpad/gerrit is probably wider than you're suggesting. My perception is that a switch would be easier than you think. That is why I brought up the idea of asking support instead of ruling things out because we perceive someone else would have an issue.

As far as I can tell only 3 people have contributed code to rpc-openstack who are not member's of our team. That doesn't suggest a great deal of investment in GitHub by other teams.

>> I don't remember seeing this apparent new release strategy announced anywhere, apart from the fact that I'd like to understand why it was changed - it can always be changed again.

>It wasn't announced. It was a casualty of our "official" early branching strategy that made the `kilo` branch something we could not cleanly or easily merge master into (a workflow that Gerrit doesn't support).

I'd like to see branching/tagging etc discussed in more detail, separately from tools.

>> The crux of it is that everyone contributing to OpenStack needs familiarity with launchpad/gerrit so why not just piggyback on that instead of having a completely separate tool chain for rpc-os when it provides us with no obvious benefit.

>Why do they need familiarity with it though? How much do they actually contribute to Nova, Neutron, Glance, etc.?

It's my experience that working with OpenStack, especially troubleshooting, means you'll end up needing to read some bug report or review. You don't need to contribute code to OpenStack to benefit from understanding launchpad/gerrit. It is worth remembering that openstack-ansible uses these tools.

It also lowers the barrier of entry for those looking to expand their skill set.

>Why is there no obvious benefit from using a service (like GitHub) that has tighter and far more reliable integrations than review.o.o + launchpad?

I haven't seen any mention of the benefits of GitHub beyond it being the status quo, but maybe I missed that bit.

>> seeing that we provide OpenStack training there will likely be a whole load of existing material that can be used to assist support instead of the training team needing to produce something new.

>Our OpenStack training does not teach you how to contribute. It teaches you how to stand-up a cloud and use it. Also, presently, we don't need training for this project because there's sufficient training documentation provided by GitHub and the internet because it's such a common workflow.

>If you're anticipating our other teams making significant contributions to OpenStack services, they'll also be making significant contributions to OpenStack's dependencies as well which are hosted on:

>- GitHub
- BitBucket
- GitLab
- And many other places

I never said there would be a sudden flood of OpenStack patches or even that using launchpad/gerrit would result in a single patch from another team. Even it this did encourage development it doesn't follow that they would start developing on projects on which OpenStack depends.

>If your only argument is that switching to those services and tools will provide them with the training they need to do this thing that they have in the past adamantly proclaimed as not their job, we shouldn't only be looking at review.o.o (gerrit) + launchpad.

The point about launchpad/gerrit is that even if you don't write code you will still end up using them if you work with OpenStack. I'm not saying it is the most important reason or the only one but I do think it's worth consideration.

byronmccollum commented:
I'd be in favor of Gerrit if we get to use infra-ci for gate testing of RPC.

BjoernT commented:
+1 the closer we are in terms of the process compared to OSA the better. I see primarily just the need for more stable branches, e.g. release branches and not single stable/kilo rather v11.0 or v11.1

byronmccollum commented:
So there's this http://gerrithub.io, which gets us the Gerrit workflow, but none of the os-infra benefits...

sigmavirus24 commented:
@byronmccollum so today we had a US/UK Engineering video conference and @dolph pointed out that upstream would most likely *not* be okay with us using their nodepool infrastructure. For an example, Fuel uses their own CI infrastructure because it is specifically a Mirantis product.

sigmavirus24 commented:
I've added 80a0f3c to split the WORKFLOW document into two separate documents - one for development, and one for releases.

You'll notice that I've removed the section about having separate branches for minor versions. I know that Support wants this kind of flexibility but QE has concerns about the complexity that it will add to our testing matrix and concerns were voiced about how long each branch lives and whether upgrades to the next major release are supported from each of those branches. Given the clear and obvious benefits that having minor branches provides to Support and customers, we should attempt to resolve the issues we have with those branches *in my opinion*. **I do not think resolving those should be a requirement for revising our workflow. I do think we should attempt to iteratively refine our workflow and that solving those problems and adding that support can be added later.**

prometheanfire commented on DEVELOPMENT_WORKFLOW.rst:
```diff
@@ -0,0 +1,138 @@
+Pre-Workflow
+------------
+
+#. Add documentation to ``master`` that is easily discoverable detailing the
```
when will we do this? 'soon'?

sigmavirus24 replied:
@d34dh0r53 I'm not sure if master is still focused on kilo or if we're moving focus to liberty. If you can answer that, I can create a document in a separate PR to do this (or as part of this PR).

prometheanfire commented on DEVELOPMENT_WORKFLOW.rst:
```diff
@@ -0,0 +1,138 @@
+   fix to a branch on |rpcos|. [#]_
+
+#. A PR is created from the branch on |rpcos|, reviewed, iterated upon and
+   eventually merged.
```
what is the default target branch? I imagine it's dependant on what master means?

sigmavirus24 replied:
I'm not sure what you're trying to ask so if you have opinions please push a revision clarifying this (preferably with more than 7 words ;)).

prometheanfire replied:
ya, think this has been clarified

sigmavirus24 replied:
Huh?

prometheanfire replied:
I thought on it more and don't care about this line of comments anymore

prometheanfire commented on DEVELOPMENT_WORKFLOW.rst:
```diff
@@ -0,0 +1,138 @@
+   eventually merged.
+
+#. Bugs meeting backport criteria [#]_ are backported to the previous branches
+   as appropriate.
```
backported using chery-picks?

sigmavirus24 replied:
Yeah, I'll provide the command as an example of how we prefer cherry-picks produced.

prometheanfire commented on DEVELOPMENT_WORKFLOW.rst:
```diff
@@ -0,0 +1,138 @@
+    It is suggested that developers use succinct branch names, e.g.,
+    ``bug/103`` or ``feature/210``, which are descriptive but also point at
+    the item being worked on. Backport PRs might consider branch names like
+    ``bug/{{backport-branch}}/103`` (e.g., ``bug/kilo/103``).
```
I'd suggest using this system for bug naming as I think it's easier to organize.

category/id/variant
bug/103/kilo

sigmavirus24 replied:
If we aren't diligent in deleting the original `bug/103` then `bug/103/kilo` will cause problems with git. `bug/kilo/103` will not be problematic in the slightest.

sigmavirus24 commented:
I'm going to move the release workflow documentation into its own PR so we can merge the developer workflow for now.

git-harry commented on RELEASE_WORKFLOW.rst:
```diff
@@ -0,0 +1,101 @@
+
+#. Each tag that is created should be tested in depth by QE and/or Support for
+   acceptance and confirmation that the acceptance criteria were met. **This
+   is designed to create shorter feedback loops.**
```
This workflow implies an increase in the workload for QE/support. Has there been any discussion as to there willingness to accept this additional work and what are the responsibilities of the two testing groups?

git-harry commented on RELEASE_WORKFLOW.rst:
```diff
@@ -0,0 +1,101 @@
+--------
+
+#. Each time a large feature targetted for the next minor release is merged, a
+   Beta tag is created, e.g., ``r11.1.0b1``, ``r11.1.0b2``, etc.
```
I agree that shorter feedback loops is something we should aim for. An alternative approach would be to have more releases. More frequent releases has the benefits of the beta tags proposed while at the same time getting the code to the customer sooner. It also makes each individual release more manageable.

prometheanfire replied:
so, in short, 'you agree'?

git-harry replied:
Yes to shorter feedback loops, no to beta tags. If the beta tag is going to get loads of testing, as described below, I don't see why it doesn't just get released.

prometheanfire replied:
because QE neeeds to ack before release.  If we do a release (because it's going to get loads of testing) and it fails tests, what do we do?  We've already released.

git-harry replied:
By release I don't just mean throw it over the wall. Line 80 [1] says each tag should be tested in depth, if beta tags are being tested in depth why not make them rc tags and release more frequently.

[1] https://github.com/rcbops/rpc-openstack/pull/667/files#diff-b567fbc3aba97a798fa3be12cc64f078R80

prometheanfire replied:
What's in a name?  That which we call a rose by any other name would smell as sweet.

git-harry commented on RELEASE_WORKFLOW.rst:
```diff
@@ -0,0 +1,101 @@
+    Ideally, they should only happen when stakeholders downstream of |rpcos|
+    request them.
+
+Workflow
```
The workflow doesn't appear to address major releases.

prometheanfire replied:
I think the current opinion is that major releases will always be messy...

git-harry commented on RELEASE_WORKFLOW.rst:
```diff
@@ -0,0 +1,101 @@
+    A patch release should only include bug fixes that should have no upgrade
+    impatch from patch release to patch release. SHA bumps of OpenStack
+    services or OpenStack Ansible should be made sparingly in patch releases.
+    Ideally, they should only happen when stakeholders downstream of |rpcos|
```
Who are those stakeholders? Is the suggestion here that the SHA bumps shouldn't be automatic for patch releases and that they should only be done in response to specific bugs that have been logged against rpc-openstack?
Most of the code lives outside this repo, I would have thought that was going to be a fairly common occurrence for a patch release.

git-harry commented on RELEASE_WORKFLOW.rst:
```diff
@@ -0,0 +1,101 @@
+
+Patch, e.g., ``r11.0.1``, ``r11.0.2``, ``r11.1.1``, etc.
+
+    A patch release should only include bug fixes that should have no upgrade
```
How are you defining upgrade impact?
